relay-server/src/utils/managed_envelope.rs

```diff
@@ -91,26 +109,66 @@ impl EnvelopeContext {
     /// As opposed to [`new`](Self::new), this does not require a queue permit. This makes it
     /// suitable for unit testing internals of the processing pipeline.
     #[cfg(test)]
-    pub fn standalone(envelope: &Envelope) -> Self {
+    pub fn standalone(envelope: Box<Envelope>) -> Self {
         Self::new_internal(envelope, None)
     }
 
-    /// Computes an envelope context from the given envelope and binds it to the processing queue.
+    /// Computes an managed envelope from the given envelope and binds it to the processing queue.
     ///
-    /// To provide additional scoping, use [`EnvelopeContext::scope`].
-    pub fn new(envelope: &Envelope, slot: SemaphorePermit) -> Self {
+    /// To provide additional scoping, use [`ManagedEnvelope::scope`].
+    pub fn new(envelope: Box<Envelope>, slot: SemaphorePermit) -> Self {
         Self::new_internal(envelope, Some(slot))
     }
 
-    /// Update the context with new envelope information.
+    /// Returns a reference to the contained [`Envelope`].
+    pub fn envelope(&self) -> &Envelope {
+        self.envelope.as_ref()
+    }
+
+    /// Returns a mutable reference to the contained [`Envelope`].
+    pub fn envelope_mut(&mut self) -> &mut Envelope {
+        self.envelope.as_mut()
+    }
+
+    /// Take the envelope out of the context and replace it with a dummy.
+    ///
+    /// Note that after taking out the envelope, the envelope summary is incorrect.
+    pub(crate) fn take_envelope(&mut self) -> Box<Envelope> {
+        Box::new(self.envelope.take_items())
+    }
+
+    /// Update the context with envelope information.
     ///
     /// This updates the item summary as well as the event id.
-    pub fn update(&mut self, envelope: &Envelope) -> &mut Self {
-        self.event_id = envelope.event_id();
-        self.summary = EnvelopeSummary::compute(envelope);
+    pub fn update(&mut self) -> &mut Self {
+        self.event_id = self.envelope().event_id();
+        self.summary = EnvelopeSummary::compute(self.envelope());
         self
     }
 
+    /// Retains only the items specified by the predicate.
+    ///
+    /// In other words, remove all elements where `f(&item)` returns `false`. This method operates
+    /// in place and preserves the order of the retained items.
```

Comment:
Nice wording of docs, I see what you did here 😄

Reply:
Not sure if it changed since my last review or if I missed it, but the predicate `retain_items` expects returns a `RetainItem` enum, and docstrings mention `false`. Could we update them?